### ISSUE TITLE
setsockopt does not work to set subscriber filters with std::string

### ISSUE THREAD

**Issue by mquirin** [zeromq/cppzmq#72]
It seems it is not possible to set subscriber filters using socket_t::setsockopt()

Expected usage would be:

    std::string filter = "meta";
    socket.setsockopt(ZMQ_SUBSCRIBE, filter); // DOES NOT WORK

Is there a workaround? 

Full example here:
https://gist.github.com/mquirin/e50dce82b6d3575351ef

PS: I am using VS 2013

**Comment by K0n63n** [comment]
This will work:
socket.setsockopt(ZMQ_SUBSCRIBE, filter.c_str(), filter.length());

**Comment by bgolenko** [comment]
I believe that template version of setsockopt function is completely broken and should probably be removed from API:

template<typename T> void setsockopt(int option_, T const& optval) {
            setsockopt(option_, &optval, sizeof(T) );
}

I can't find even a single reasonable use case where it would behave as intended.

**Comment by K0n63n** [comment]
It works fine for all int type socket options like:

```
socket.setsockopt(ZMQ_IMMEDIATE, 1);
socket.setsockopt(ZMQ_LINGER, 1000);
```

But I guess there should have been an overload for the string case:

```
void setsockopt(int option_, const string& optval)
{
    setsockopt(option_, optval.c_str(), optval.length() );
}
```

**Comment by bgolenko** [comment]
You are right, it will behave correctly for integer (assuming int size does not matter here). However its just one single use case.
Will it behave correctly if I pass int16_t or int64_t ? Regardless of system endiannes ?

I think it would be safest to replace this template with two functions:
    
    // your overload for std::string - will handle old C-style strings as well
    void setsockopt(int option_, const string& optval)
    {
        setsockopt(option_, optval.c_str(), optval.length() );
    }

    void setsockopt(int option_, int optval)
    {
        setsockopt(option_, &optval, sizeof(int) );
    }

**Comment by keith-dev** [comment]
You'd expect to pass an integral value there, not string. I mean, what't in that string? What if it's empty or contains a leading whitespace character?

**Comment by bgolenko** [comment]
See original question. For ZMQ PUB/SUB you need to pass filter string:
```c++    
    socket.setsockopt(ZMQ_SUBSCRIBE, "myfilter", strlen("myfilter"));
```
This is misleading, because if you check Python examples you may do something like this:
```c++    
socket.setsockopt(ZMQ_SUBSCRIBE, "myfilter");
```
It works in Python, but for C++ its translated into the call
```c++    
    socket.setsockopt(ZMQ_SUBSCRIBE, "myfilter", sizeof(char)); 
```
or if you use std::string
```c++    
    socket.setsockopt(ZMQ_SUBSCRIBE, std::string("myfilter"), sizeof(std::string));  
```
In both cases its invalid and subscriber will fail to work

**Comment by K0n63n** [comment]
@bgolenko: It works fine for both int and int64_t, hence the templated function. There are currently no socket options with parameter type of int16_t as I can see from documentation: http://api.zeromq.org/4-1:zmq-setsockopt. However if there were any int16_t or any other fixed-size-type parameters, it would work for those as well.

@keith-dev There are several socket options that take a c-string as input.

**Comment by keith-dev** [comment]
If you think it's useful, you could create a suitable specialized template. If it proved reliable, you could submit it to the project.

**Comment by phlptp** [comment]
If it is any help I put this code before the setsockopt template for use with strings.  I haven't rigorously tested this code but it seems to work for the uses I have for it, namely setting subscriptions
```
inline void setsockopt(int option_, const std::string &optval)
{
	if (optval.empty())
	{
		throw error_t();
	}
	setsockopt(option_, optval.c_str(), optval.length());
}
```

**Comment by keith-dev** [comment]
It shouldn't be an error to pass an empty string.  That's the only way to subscribe to everything.
